src/main/java/com/ge/predix/uaa/token/lib/FastTokenServices.java

```diff
@@ -128,8 +128,23 @@ public OAuth2Authentication loadAuthentication(final String accessToken) throws
 
         verifyIssuer(iss);
 
+        /* Memory leak was identified in performance tests and defect is added with id DE68869. This check is added to
+        address a memory leak possible through lazy expiration mechanism in passive expiring map.
+        Though we are initializing passive expiring map with time to expire the entries. The entries are not flushed
+        unless there is a get call made on the expired entries.
+
+        For instance a loadAuthentication call made it adds the signature verifier for that access token in tokenKeys
+        map this entry is not flushed if there is no get call made with the same access token after the expired time.
+        This could lead to memory leak when get calls on expired tokens does not happen.
+
+        This fastTokenMap.isEmpty() call scans the map and flushes all the expired entries.
+         */
+        SignatureVerifier verifier = null;
+        if (!this.tokenKeys.isEmpty()) {
```

Comment:
Discussed this with @metacubed and @sanjeevchopra and we should consider switching to Caffeine which works well with Spring and provides weak keys, soft values and an expiration-after-write policy: https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-caching.html#boot-features-caching-provider-caffeine .

Reply:
I will look into caffeine. But as we tried earlier today with google cache weak keys and soft values with expiration after write policy is not providing a desired behavior. I think weak keys, weak values and expiration after write policy should be considered in caffeine as well.

Reply:
If we can verify that the cache doesn’t expire the entry before the duration set by the policy with weak values then yes it should be fine to use.

Reply:
Caffeine in general has daemons performing cleanup tasks kicked off during any access and hence do not require explicit calls to isEmpty() or size() to perform expired evictions.

As per using weak keys and soft values, the use case needs to be vetted out as such the entry is not expired before the ttl policy.

Reply:
FYI, the daemon (FJP.commonPool()) can be disabled if desired, using `Caffeine.executor(Runnable::run)`. The internal algorithms are all amortized O(1), so the goal is provide more consistent latencies when user-facing.

Internally expiration uses either (a) LRU lists if a fixed duration or (b) Timer Wheels if variable. It looks like you have a fixed TTL, so it merely has to reorder the entry to the back of a doubly-linked list on a write. Expiring is then peeking at the front of the list and pruning as needed.

Please note that a weak keyed cache will lookup by identity, not equality, which may not be what you want. (`cache.put("a", "b")` will not be found by `cache.get(new String("a")`)

P.S. Hi Sanjeev!

Reply:
Hey Ben! ...small world :) 
(@ben-manes )